src/validation.cpp

```diff
@@ -3649,7 +3673,28 @@ bool BlockManager::LoadBlockIndex(
         if (pindex->IsAssumedValid() ||
                 (pindex->IsValid(BLOCK_VALID_TRANSACTIONS) &&
                  (pindex->HaveTxsDownloaded() || pindex->pprev == nullptr))) {
-            block_index_candidates.insert(pindex);
+
+            // Fill each chainstate's block candidate set. Add all blocks as
+            // candidates if the chainstate is allowed to rely on assumed-valid
+            // blocks. Otherwise avoid adding assumed-valid blocks so that
+            // non-assumed-valid chainstates will download and validate all earlier,
+            // unvalidated blocks.
+            //
+            // Note: This is considering all blocks whose height is greater or
+            // equal to the first assumed-valid block to be assumed-valid
+            // blocks, and excluding them from the background chainstate's
+            // setBlockIndexCandidates set. This does mean that some blocks
+            // which are not technically assumed-valid (later blocks on a fork
+            // beginning before the first assumed-valid block) might not get
+            // added to the the background chainstate, but this is ok, because
+            // they will still be attached to the active chainstate if they
+            // actually contain more work.
```

Comment:
in commit 36e2bc4275fdb0d72390215fcea4d499f8049701:

I don't understand why it is safe to not attach them. If the assumed valid block is sufficiently old, it may cost little POW to create an alternative fork up to that height, and confusing the background chainstate that it can't connect to the active chainstate?

What is the reason for the if-condition in the first place? What would happen if both candidate sets are fully populated?

Reply:
If both candidate sets are fully populated, when ActivateBestChain is called on the background validation chainstate, [`FindMostWorkChain()`](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2410-L2416) will return the assumed-valid block index entry as installed by snapshot activation, and it won't think it has to enter IBD to download and fully validate the chain up to the snapshot base block. I'll include this in the commentary above since people (myself included) seem to constantly forget why we need this conditional.

Reply:
> If the assumed valid block is sufficiently old, it may cost little POW to create an alternative fork up to that height, and confusing the background chainstate that it can't connect to the active chainstate?

I'm confused by this because the background chainstate shouldn't be considering forks anyway; its purpose is to verify the assumed-valid blocks beneath the snapshot and it does not handle reorgs or forks.

Reply:
Ok, I guess I am missing context on how the download logic will be modified for assumevalid.

If the assumevalid chain is assumed to be always linear, then maybe a check that all assumed blocks are parents of the assumed tip block makes sense?